numba/inline_closurecall.py

```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
     Return return statement by assigning directly to target, and a jump.
     """
     for label, block in blocks.items():
+        casts = []
         for i in range(len(block.body)):
             stmt = block.body[i]
             if isinstance(stmt, ir.Return):
                 assert(i + 1 == len(block.body))
                 block.body[i] = ir.Assign(stmt.value, target, stmt.loc)
                 block.body.append(ir.Jump(return_label, stmt.loc))
+                # remove cast of the returned value
+                for cast in casts:
+                    if cast.target.name == stmt.value.name:
+                        cast.value = cast.value.value
+            elif isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr) and stmt.value.op == 'cast':
+                casts.append(stmt)
+
 
-def _add_definition(func_ir, block):
+def _add_definitions(func_ir, block):
     """
-    Add variable definitions to parent func_ir
+    Add variable definitions found in a block to parent func_ir.
     """
     definitions = func_ir._definitions
     assigns = block.find_insts(ir.Assign)
     for stmt in assigns:
         definitions[stmt.target.name].append(stmt.value)
 
+class GuardException(Exception):
+    pass
+
+def require(cond):
+    """
+    Raise GuardException if the given condition is False.
+    """
+    if not cond:
+       raise GuardException
+
+def guard(func, *args):
```

Comment:
Not sure if `contextmanager` can help to achieve the same effect. The point of using `require` and `guard` (in many of the code that follows) is to refactor [nested if-statements](https://softwareengineering.stackexchange.com/questions/47789/how-would-you-refactor-nested-if-statements).

Typical solution (like mentioned in the above link) is using `if (!cond) { return None }`. But this has a few problems of its own: (1) each check becomes statement, and hence cannot be composed together like expressions; (2) the return value `None` still has to be manually checked by the caller; (3) `return None` doesn't play well with multiple return values. 

So the idiom of `require` and `guard` I use here is to convert between exceptions and optional values (with the presence of None). Similar tricks were done in functional programming languages like Haskell in the form of `Maybe` monad to ease programming in this kind of control flow. But Python doesn't have the same level of abstraction power, so I use exception to manipulate the control flow to achieve a similar effect.

Certainly I'm not the first one who invented this trick, I've seen it being used by other people. I wish I could give more pointers, but I don't exactly have them.